lib/Mojo/WebSocketProxy/Backend/ConsumerGroups.pm

```diff
@@ -0,0 +1,347 @@
+
+=item * C<before_get_rpc_response>  arrayref of subroutines to be run before the remote response is received, with args: C<< $c >> and C<< req_storage >>
+
+=item * C<after_get_rpc_response> arrayref of subroutines to be run after the remote response is received; called with args: C<< $c >> and C<< req_storage >>
+called only when there is an response from the remote call.
+
+=item * C<before_call> arrayref of subroutines called before the request to the remote service is made.
+
+=item * C<rpc_failure_cb> a subroutine reference to call if the remote call fails. Called with C<< Mojolicious::Controller >>, the rpc_response and C<< $req_storage >>
+
+=back
+
+=back
+
+Returns undef.
+
+=cut
+
+sub call_rpc {
+    my ($self, $c, $req_storage) = @_;
+
+    my ($msg_type, $request_data) = $self->_prepare_request_data($c, $req_storage);
+
+    my $rpc_response_cb = $self->get_rpc_response_cb($c, $req_storage);
+    my $before_get_rpc_response_hooks = delete($req_storage->{before_get_rpc_response}) || [];
+    my $after_got_rpc_response_hooks  = delete($req_storage->{after_got_rpc_response})  || [];
+    my $before_call_hooks             = delete($req_storage->{before_call})             || [];
+    my $rpc_failure_cb                = delete($req_storage->{rpc_failure_cb});
+
+    foreach my $hook ($before_call_hooks->@*) { $hook->($c, $req_storage) }
+
+    $self->request($request_data)->then(
+        sub {
+            my ($message) = @_;
+
+            foreach my $hook ($before_get_rpc_response_hooks->@*) { $hook->($c, $req_storage) }
+
+            return Future->done unless $c && $c->tx;
+            my $api_response;
+            my $result;
+
+            try {
+                $result = MojoX::JSON::RPC::Client::ReturnObject->new(rpc_response => $message);
+                foreach my $hook ($after_got_rpc_response_hooks->@*) { $hook->($c, $req_storage, $result) }
+                $api_response = $rpc_response_cb->($result->result);
+            }
+            catch {
+                my $error = $@;
+                $rpc_failure_cb->($c, $result, $req_storage) if $rpc_failure_cb;
+                $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+            };
+
+            $c->send({json => $api_response}, $req_storage);
+
+            return Future->done;
+        }
+        )->catch(
+        sub {
+            my $error = shift;
+            my $api_response;
+
+            return Future->done unless $c && $c->tx;
+
+            if ($error eq 'Timeout') {
+                $api_response = $c->wsp_error($msg_type, 'RequestTimeout', 'Request is timed out.');
+            } else {
+                $api_response = $c->wsp_error($msg_type, 'WrongResponse', 'Sorry, an error occurred while processing your request.');
+            }
+            $rpc_failure_cb->($c, undef, $req_storage) if $rpc_failure_cb;
```

Comment:
not passing message

Reply:
I tried to be consistent with job async, second argument should be rpc response, but in case of time out, it'll be undef. https://github.com/binary-com/perl-Mojo-WebSocketProxy/blob/master/lib/Mojo/WebSocketProxy/Backend/JobAsync.pm#L205